[wallet] [rpc] sendtoaddress/sendmany: Add explicit feerate option

kallewoof opened the pull request:
This lets users pick their own fees when using `sendtoaddress`/`sendmany` if they prefer this over the estimators.

jonasschnelli commented:
Not opposed to add that. Though IMO the more experience users (which are those who may use explicit fee-rates) should use create/fund/sendrawtx.

Concept ACK

promag commented on src/policy/fees.cpp:
```diff
@@ -54,6 +54,7 @@ bool FeeModeFromString(const std::string& mode_string, FeeEstimateMode& fee_esti
         {"UNSET", FeeEstimateMode::UNSET},
         {"ECONOMICAL", FeeEstimateMode::ECONOMICAL},
         {"CONSERVATIVE", FeeEstimateMode::CONSERVATIVE},
+        {"EXPLICIT", FeeEstimateMode::EXPLICIT},
```
The intent sounds more like no estimation so `{"NONE", FeeEstimateMode::NONE}`?

kallewoof replied:
I don't know. People could mistake that to mean "no fee".

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -445,12 +445,15 @@ UniValue sendtoaddress(const JSONRPCRequest& request)
             "       \"ECONOMICAL\"\n"
             "       \"CONSERVATIVE\"\n"
+            "       \"EXPLICIT\"\n"
+            "9. feerate                (numeric, optional) Fee rate (sat/kb) to use for EXPLICIT mode.\n"
```
Mention that `estimate_mode` must be set to (currenlty) `EXPLICIT` to use this value?

kallewoof replied:
`to use for EXPLICIT mode` is not enough?

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -494,6 +497,12 @@ UniValue sendtoaddress(const JSONRPCRequest& request)
 
+    if (!request.params[8].isNull()) {
+        if (coin_control.m_fee_mode != FeeEstimateMode::EXPLICIT) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Explicit fee rate can only be used when estimate mode is set to EXPLICIT");
```
Add test for error.

promag reviewed (commented):
Please add tests.

esotericnonsense commented:
Concept ACK. I don't like that this results in sendtoaddress having 8 parameters but I don't see an immediately obvious way to avoid that.

Sendmany could get it too?

kallewoof commented:
Can't fix people putting too high fees in, but if they put too low, they should be able to bump the fee unless they specifically disable it. As such, RBF is default-on when fees are explicit (3d14d35).

Added tests (1768936).

MeshCollider commented:
Re: the 8 parameter issue @esotericnonsense mentioned, it would be cleaner to convert most parameters to a JSON object input like `bumpfee`, `listunspent`, etc. use since most are independent of each other and optional

kallewoof commented on src/rpc/client.cpp:
```diff
@@ -39,6 +39,7 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "sendtoaddress", 4, "subtractfeefromamount" },
     { "sendtoaddress", 5 , "replaceable" },
     { "sendtoaddress", 6 , "conf_target" },
+    { "sendtoaddress", 8, "feerate"},
```
Self-nit: space after `"feerate"`

luke-jr reviewed (changes requested):
Concept ACK, but `sendmany` should get it too.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -445,12 +445,15 @@ UniValue sendtoaddress(const JSONRPCRequest& request)
             "       \"ECONOMICAL\"\n"
             "       \"CONSERVATIVE\"\n"
+            "       \"EXPLICIT\"\n"
+            "9. feerate                (numeric, optional) Fee rate (sat/kb) to use for EXPLICIT mode. (Note: replaceable defaults to true when used)\n"
```
This isn't in the argument list above.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -494,6 +497,14 @@ UniValue sendtoaddress(const JSONRPCRequest& request)
         }
     }
 
+    if (!request.params[8].isNull()) {
```
It doesn't make sense to add a new param for this. Instead, make `params[6]` be conf target *or* feerate...

kallewoof commented:
Addressed @luke-jr requests (double-use `conf_target` and add feature to `sendmany`).

practicalswift commented on src/wallet/rpcwallet.cpp:
```diff
@@ -402,16 +405,21 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
+        } else {
+            coin_control.m_feerate = CFeeRate(request.params[6].get_int());
+            // default RBF to true for explicit fee rate mode
+            coin_control.m_signal_bip125_rbf = (coin_control.m_signal_bip125_rbf ? *coin_control.m_signal_bip125_rbf : false) | request.params[5].isNull();
```
```
2018-09-19 13:23:27 clang-tidy(pr=11413): src/wallet/rpcwallet.cpp:420:48: warning: implicit conversion bool -> 'int' [readability-implicit-bool-conversion]
2018-09-19 13:23:27 clang-tidy(pr=11413): src/wallet/rpcwallet.cpp:420:129: warning: implicit conversion bool -> 'int' [readability-implicit-bool-conversion]
```

kallewoof replied:
Thanks. I assume this is referring to the `|`?

practicalswift commented on src/wallet/rpcwallet.cpp:
```diff
@@ -833,16 +842,22 @@ static UniValue sendmany(const JSONRPCRequest& request)
+        } else {
+            coin_control.m_feerate = CFeeRate(request.params[6].get_int());
+            // default RBF to true for explicit fee rate mode
+            coin_control.m_signal_bip125_rbf = (coin_control.m_signal_bip125_rbf ? *coin_control.m_signal_bip125_rbf : false) | request.params[5].isNull();
```
```
2018-09-19 13:23:27 clang-tidy(pr=11413): src/wallet/rpcwallet.cpp:857:48: warning: implicit conversion bool -> 'int' [readability-implicit-bool-conversion]
2018-09-19 13:23:27 clang-tidy(pr=11413): src/wallet/rpcwallet.cpp:857:129: warning: implicit conversion bool -> 'int' [readability-implicit-bool-conversion]
```

sipa commented:
Concept ACK

MeshCollider commented:
utACK https://github.com/bitcoin/bitcoin/pull/11413/commits/d9b1c428486b8f0acb92825a01d7dc5c0994ca24, my earlier concern about adding another parameter is no longer relevant :)

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -355,17 +355,20 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
             "6. replaceable            (boolean, optional) Allow this transaction to be replaced by a transaction with higher fees via BIP 125\n"
-            "7. conf_target            (numeric, optional) Confirmation target (in blocks)\n"
+            "                             Defaults to false unless explicit estimate mode is used, in which case it defaults to true\n"
+            "7. conf_target            (numeric, optional) Confirmation target (in blocks), or fee rate (sat/kb) if explicit estimate mode is used\n"
```
Fee rates are passed in as BTC/kB (see settxfee for example) and amounts can be passed in as either num or string.

kallewoof replied:
Fixed.

promag reviewed (commented):
utACK, needs squash and could add release notes.

kallewoof committed:
doc: add release notes for explicit fee estimators and bumpfee change

promag commented on doc/release-notes-11413.md:
```diff
@@ -0,0 +1,5 @@
+Low-level RPC changes
```
nit, I may be wrong but I think you could increment `doc/release-notes-14454.md`?

kallewoof replied:
Hm. Does it matter? I assume the files are merged together at the end anyway.

kallewoof commented:
@promag Added release notes & squashed. Also fixed sendmany, which I had forgotten.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -357,17 +357,20 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
             "6. replaceable            (boolean, optional) Allow this transaction to be replaced by a transaction with higher fees via BIP 125\n"
-            "7. conf_target            (numeric, optional) Confirmation target (in blocks)\n"
+            "                             Defaults to false unless explicit estimate mode is used, in which case it defaults to true\n"
+            "7. conf_target            (numeric, optional) Confirmation target (in blocks), or fee rate (BTC/kb) if explicit estimate mode is used\n"
```
Per kB, not per kb[it].

luke-jr replied:
Also should use `CURRENCY_UNIT` here.

kallewoof commented:
@luke-jr Thanks for review! I switched to `" + CURRENCY_UNIT + "/kB` as suggested.

MeshCollider commented on src/wallet/rpcwallet.cpp:
```diff
@@ -377,6 +378,7 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
             + HelpExampleCli("sendtoaddress", "\"1M72Sfpbz1BPpXFHz9m3CdqATR44Jvaydd\" 0.1")
             + HelpExampleCli("sendtoaddress", "\"1M72Sfpbz1BPpXFHz9m3CdqATR44Jvaydd\" 0.1 \"donation\" \"seans outpost\"")
             + HelpExampleCli("sendtoaddress", "\"1M72Sfpbz1BPpXFHz9m3CdqATR44Jvaydd\" 0.1 \"\" \"\" true")
+            + HelpExampleCli("sendtoaddress", "\"1M72Sfpbz1BPpXFHz9m3CdqATR44Jvaydd\" 0.1 \"\" \"\" false true 2000 EXPLICIT")
```
This example feerate needs updating now to be BTC/kB, 2000 BTC/kB is probably a bit high 😄

meshcollider reviewed (commented):
re-utACK https://github.com/bitcoin/bitcoin/pull/11413/commits/7344398bb5e00f242cc2c0adcbdc1463f6ebab7f

luke-jr commented on test/functional/wallet_basic.py:
```diff
@@ -22,6 +22,8 @@ class WalletTest(BitcoinTestFramework):
         self.num_nodes = 4
         self.setup_clean_chain = True
+        # node 2 needs a txindex to look up raw transactions
+        self.extra_args = [[], [], ["-txindex"], []]
```
This is part of the wrong commit. The `sendtoaddress` test only calls `getrawtransaction` on a transaction in the mempool, which doesn't need a tx index.

Also, the `sendmany` test that relies on this *could* just pass the block hash explicitly to avoid building the tx index.

luke-jr replied:
Actually, you already removed the `getrawtransaction` from sendmany, so this seems completely unnecessary?

kallewoof replied:
Didn't realize that, thanks. Removed these 2 lines!

luke-jr commented on test/functional/wallet_basic.py:
```diff
@@ -336,6 +351,25 @@ def run_test(self):
+        )
+        tx_size = self.get_vsize(self.nodes[2].gettransaction(txid)['hex'])
+        self.sync_all([self.nodes[0:3]])
+        self.nodes[0].generate(1)
```
Since the block is mined by node 0, and the balance is on node 2, you need another `sync_all` here or else there's a race condition.

kallewoof replied:
@luke-jr Thanks, fixed!

Sjors commented:
Concept ACK. Alernatively you could replace `EXPLICIT` with `btc/kb`. That's more readable, prevents accidents when people forget the unit account isn't satoshi per byte (see #16257) and makes it easy to add `sat/b` as a unit a later.

kallewoof commented:
@Sjors I love it. Will definitely do, thanks!

kallewoof commented:
Done. Also squashed things.

For shitcoiners, thisis a drawback as a new hard-coded coin name ("btc" in "btc/kb") is introduced, but don't think anyone cares about that.

Sjors commented:
Ah, I missed this update, here was my stab at the same, plus satoshi per byte: https://github.com/Sjors/bitcoin/commit/a0996c125d59e9809306818533eba5cc6e512eb8

MarcoFalke commented on src/util/fees.cpp:
```diff
@@ -31,6 +31,7 @@ bool FeeModeFromString(const std::string& mode_string, FeeEstimateMode& fee_esti
         {"UNSET", FeeEstimateMode::UNSET},
         {"ECONOMICAL", FeeEstimateMode::ECONOMICAL},
         {"CONSERVATIVE", FeeEstimateMode::CONSERVATIVE},
+        {"btc/kb", FeeEstimateMode::BTC_KB},
```
I think we use upper case. See also

```
src/policy/feerate.cpp:const std::string CURRENCY_UNIT = "BTC";
```

which you can use to express your love of shitcoins.

kallewoof replied:
I am very tempted to make the parser case insensitive and use CURRENCY_UNIT as you say. Otherwise it's a bit of a pain to require users to type "BTC/kb" every time (as opposed to just "btc/kb")

MarcoFalke replied:
Also, there is a difference between kilobit (kb) and kilobyte (kB)

Sjors replied:
+1 for case insensitive

Also fine with CURRENCY_UNIT

I don't care about the `kB` vs `kb` distinction as far as the parser goes, but documentation wise we should do it correctly. Users will out by themselves it's case insensitive :-)

luke-jr replied:
RPC is for software, not users...

kallewoof replied:
It's for any entity that doesn't use the GUI, which includes non-software me.

kallewoof committed:
added CURRENCY_ATOM to express minimum indivisible unit

also moved CURRENCY_* into feerate.h file to work around MSVC bug

kallewoof commented:
I ended up refactoring the uppercase/lowercase code quite heavily, which may deserve its own PR. Feedback on this welcome.

Sjors commented on src/policy/fees.h:
```diff
@@ -50,6 +50,8 @@ enum class FeeEstimateMode {
     ECONOMICAL,   //!< Force estimateSmartFee to use non-conservative estimates
     CONSERVATIVE, //!< Force estimateSmartFee to use conservative estimates
+    BTC_KB,       //!< Use explicit BTC/kB fee given in coin control
+    SAT_B,        //!< Use explicit SAT/b fee given in coin control
```
Should be upper case B.

Sjors commented on src/wallet/rpcwallet.cpp:
```diff
@@ -360,11 +376,13 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
                     {"replaceable", RPCArg::Type::BOOL, /* default */ "wallet default", "Allow this transaction to be replaced by a transaction with higher fees via BIP 125"},
-                    {"conf_target", RPCArg::Type::NUM, /* default */ "wallet default", "Confirmation target (in blocks)"},
-                    {"estimate_mode", RPCArg::Type::STR, /* default */ "UNSET", "The fee estimate mode, must be one of:\n"
+                    {"conf_target", RPCArg::Type::NUM, /* default */ "wallet default", "Confirmation target (in blocks), or fee rate (" + CURRENCY_UNIT + "/kB) if " + (CURRENCY_UNIT + "/kB") + " estimate mode is used"},
```
The second one should be CURRENCY_ATOM + "/B"

kallewoof replied:
Hm? No it is literally saying "fee rate (BTC/kB) if BTC/kB estimate mode is used". Though since we now have two, maybe it should be "fee rate (for BTC/KB or SAT/B estimate modes)" instead.

Sjors commented on src/wallet/rpcwallet.cpp:
```diff
@@ -816,11 +837,13 @@ static UniValue sendmany(const JSONRPCRequest& request)
                     {"replaceable", RPCArg::Type::BOOL, /* default */ "wallet default", "Allow this transaction to be replaced by a transaction with higher fees via BIP 125"},
-                    {"conf_target", RPCArg::Type::NUM, /* default */ "wallet default", "Confirmation target (in blocks)"},
-                    {"estimate_mode", RPCArg::Type::STR, /* default */ "UNSET", "The fee estimate mode, must be one of:\n"
+                    {"conf_target", RPCArg::Type::NUM, /* default */ "wallet default", "Confirmation target (in blocks), or fee rate (" + CURRENCY_UNIT + "/kB) if " + (CURRENCY_UNIT + "/kB") + " estimate mode is used"},
```
Same issue as above

Sjors reviewed (commented):
A separate PR for the uppercase/lowercase makes sense. It looks simple enough, but it does touch networking code.

utACK 626d26c modulo documentation bug.

kallewoof commented:
Addressed @Sjors nits. Note there is #16566 which contains the case function refactor commit only.

Sjors commented:
utACK 601da0f981078a99b3eb5d689f42c7937199a375

luke-jr commented on src/util/system.cpp:
```diff
@@ -388,7 +388,7 @@ bool ArgsManager::ParseParameters(int argc, const char* const argv[], std::strin
             key.erase(is_index);
         }
 #ifdef WIN32
-        std::transform(key.begin(), key.end(), key.begin(), ToLower);
```
This seems to clearly perform worse. Is there a reason not to keep the current way? (A std::string ToLower overload could be *in addition* to this.)

Sjors replied:
@luke-jr this bit is moved to #16566

kallewoof replied:
I could keep the in-place transform, but seriously, this is called a couple of dozen times once at startup and not again. I think we want to aim for readability on this one.

luke-jr commented on src/policy/fees.h:
```diff
@@ -50,6 +50,8 @@ enum class FeeEstimateMode {
     ECONOMICAL,   //!< Force estimateSmartFee to use non-conservative estimates
     CONSERVATIVE, //!< Force estimateSmartFee to use conservative estimates
+    BTC_KB,       //!< Use explicit BTC/kB fee given in coin control
+    SAT_B,        //!< Use explicit SAT/B fee given in coin control
```
It seems pretty ugly to pass the various units into fee estimate mode here. Can't we convert to EXPLICIT for internal stuff?

kallewoof replied:
I don't understand what you are suggesting.

luke-jr replied:
BTC/kB and sat/B should be limited to the RPC interface. Once it gets to CCoinControl, it should just be EXPLICIT.

kallewoof replied:
Ahh, but that complicates the parser part though. Right now, it can take a string and output a `FeeEstimateMode`. With your suggestion that wouldn't be sufficient.

luke-jr replied:
It may actually simplify it. There's only a single call that can be inlined here.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -413,10 +433,6 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
-    }
-
     if (!request.params[7].isNull()) {
         if (!FeeModeFromString(request.params[7].get_str(), coin_control.m_fee_mode)) {
```
Let's move this into `SetFeeEstimateParam`...?

kallewoof replied:
Done!

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -360,11 +376,13 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
                     {"replaceable", RPCArg::Type::BOOL, /* default */ "wallet default", "Allow this transaction to be replaced by a transaction with higher fees via BIP 125"},
-                    {"conf_target", RPCArg::Type::NUM, /* default */ "wallet default", "Confirmation target (in blocks)"},
-                    {"estimate_mode", RPCArg::Type::STR, /* default */ "UNSET", "The fee estimate mode, must be one of:\n"
+                    {"conf_target", RPCArg::Type::NUM, /* default */ "wallet default", "Confirmation target (in blocks), or fee rate (for " + CURRENCY_UNIT + "/KB or " + CURRENCY_ATOM + "/B estimate modes)"},
```
"KB" is 1024*1024 bytes; we should at least document it as "kB"

kallewoof replied:
Estimate modes are all uppercase, so I don't think the convention applies here.

luke-jr replied:
Since we're making them case-insensitive, it seems like a good opportunity to introduce mixed case here.

kallewoof replied:
I'm not opposed to that, though it would make the PR unnecessarily bigger.

luke-jr commented on src/policy/feerate.cpp:
```diff
@@ -8,6 +8,7 @@
 #include <tinyformat.h>
 
 const std::string CURRENCY_UNIT = "BTC";
+const std::string CURRENCY_ATOM = "SAT";
```
Isn't "sat" usually lowercase?

kallewoof replied:
Yeah, hrm. The one place it's used right now is uppercase, though, so it may make sense to just do `ToLower(CURRENCY_ATOM)` in future uses. Or alternatively I use `ToUpper(CURRENCY_ATOM)` here...

luke-jr replied:
IMO `ToUpper(CURRENCY_ATOM)` makes more sense, since it could just as well be "cXBT" or such in some hypothetical sidechain or future hardfork.

kallewoof replied:
I am inclined to agree, but a lot of `ToUpper` was necessary: 4b304fefc517404bb109041cfcfa2a2c576c9a42

luke-jr replied:
IMO only the one case in `FeeModeFromString` is actually necessary.

kallewoof replied:
You mean if we mix-case the estimate modes? Yeah.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -375,6 +393,8 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
             + HelpExampleCli("sendtoaddress", "\"1M72Sfpbz1BPpXFHz9m3CdqATR44Jvaydd\" 0.1 \"donation\" \"seans outpost\"")
             + HelpExampleCli("sendtoaddress", "\"1M72Sfpbz1BPpXFHz9m3CdqATR44Jvaydd\" 0.1 \"\" \"\" true")
+            + HelpExampleCli("sendtoaddress", "\"1M72Sfpbz1BPpXFHz9m3CdqATR44Jvaydd\" 0.1 \"\" \"\" false true 0.00002 " + (CURRENCY_UNIT + "/kB"))
+            + HelpExampleCli("sendtoaddress", "\"1M72Sfpbz1BPpXFHz9m3CdqATR44Jvaydd\" 0.1 \"\" \"\" false true 2 " + (CURRENCY_ATOM + "/b"))
```
Should be an uppercase B here.

luke-jr commented on src/util/fees.cpp:
```diff
@@ -26,15 +28,31 @@ std::string StringForFeeReason(FeeReason reason) {
 }
 
-bool FeeModeFromString(const std::string& mode_string, FeeEstimateMode& fee_estimate_mode) {
+const std::map<std::string, FeeEstimateMode>& FeeModeMap() {
```
Is there any benefit to the function? (As opposed to a global map)

kallewoof replied:
The windows build (AppVeyor) fails to generate the map if I don't use a function. :/

luke-jr commented on src/util/fees.cpp:
```diff
@@ -26,15 +28,31 @@ std::string StringForFeeReason(FeeReason reason) {
+std::string FeeModes(const std::string& delimiter = ", ") {
+    std::ostringstream r;
+    bool first = true;
+    for (const auto& i : FeeModeMap()) {
```
This ends up with the uppercased strings in docs rather than the correct "BTC/kB" and "sat/B"

kallewoof replied:
Everything is uppercase right now. I'm honestly not sure if changing that is the right approach.

kallewoof replied:
Code has been updated to show correct capitalization.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3033,20 +3043,7 @@ void FundTransaction(CWallet* const pwallet, CMutableTransaction& tx, CAmount& f
-        }
-        if (options.exists("estimate_mode")) {
-            if (options.exists("feeRate")) {
-                throw JSONRPCError(RPC_INVALID_PARAMETER, "Cannot specify both estimate_mode and feeRate");
```
This error seems to be lost?

kallewoof replied:
Good catch; re-added into the feeRate block above.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3329,10 +3330,11 @@ static UniValue bumpfee(const JSONRPCRequest& request)
             true, true);
 
-        if (options.exists("confTarget") && options.exists("totalFee")) {
+        auto conf_target_exists = options.exists("confTarget") || options.exists("conf_target");
```
It might be better to replace this with `!conf_target.IsNull()`

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3329,10 +3330,11 @@ static UniValue bumpfee(const JSONRPCRequest& request)
             true, true);
 
-        if (options.exists("confTarget") && options.exists("totalFee")) {
+        auto conf_target_exists = options.exists("confTarget") || options.exists("conf_target");
```
There's a type check above that needs to be updated for the alias. (Prefer the alias being a separate commit.)

kallewoof replied:
Split alias part into separate commit and added type check.

kallewoof committed:
rpc/wallet: add conf_target as alias to confTarget in bumpfee

kallewoof commented:
Split into 6 commits:
- CURRENCY_ATOM addition
- conf_target alias in bumpfee
- FeeMode doc helper function
- make fee estimate modes case insensitive
- add BTC/KB and SAT/B modes (all caps, like other modes)
- release notes

Sjors commented on src/wallet/rpcwallet.cpp:
```diff
@@ -158,6 +158,30 @@ static std::string LabelFromValue(const UniValue& value)
+        }
+        cc.m_feerate = CFeeRate(fee_rate);
+        // default RBF to true for explicit fee rate mode
+        cc.m_signal_bip125_rbf = (cc.m_signal_bip125_rbf ? *cc.m_signal_bip125_rbf : false) || may_set_rbf;
```
This seems unnecessary (or you need another test case). You can drop `may_set_rbf` and use `if(cc.m_signal_bip125_rbf) {*cc.m_signal_bip125_rbf = true}`. See 4594923046585453beef83e833e6a422a78b66b2

kallewoof replied:
Shouldn't that be `if (!cc.m_signal_bip125_rbf)`? I thought optionals returned `0` when unset, and we only wanna set it if it's not been set already, right?

kallewoof replied:
Wading through the slugfest that is boost, it seems that the most correct way to determine if an optional is uninitialized is to do `== boost::none`. I.e.
```C++
if (cc.m_signal_bip125_rbf == boost::none) { ... }
```

Sjors replied:
+1 for using `boost::none`

Sjors reviewed (commented):
ACK 81a42039e0664682ec8a35bbb9327acfce1837e2 modulo RBF code.

Sjors commented:
ACK c9971be689aab3242f35447ad20e187915f59418

kallewoof commented:
@Sjors Apologies, I missed your preserve-ordering suggestion. How does fa09386 look?

Sjors commented:
Looks better:
<img width="866" alt="Schermafbeelding 2019-08-16 om 11 18 43" src="https://user-images.githubusercontent.com/10217/63157662-b5e56000-c017-11e9-9a37-3ccabecf3b09.png">

kallewoof commented:
Cool! I realized looking at your output that "unset" should be "UNSET". Fixup'd.

luke-jr commented on src/util/fees.cpp:
```diff
@@ -26,15 +27,29 @@ std::string StringForFeeReason(FeeReason reason) {
 
-bool FeeModeFromString(const std::string& mode_string, FeeEstimateMode& fee_estimate_mode) {
+const std::map<std::string, FeeEstimateMode>& FeeModeMap() {
     static const std::map<std::string, FeeEstimateMode> fee_modes = {
```
... but std::map is ordered, isn't it? That's why C++ added std::unordered_map?

kallewoof replied:
@luke-jr I believe it std::set orders, which for std::string I think means they become ordered alphabetically.

luke-jr replied:
If we really care about the order, std::map lets you specify that.

kallewoof replied:
@luke-jr We want them in the order they are added. It looks like `std::map` is not trivial to do this?

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -356,11 +380,10 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
-            "       \"ECONOMICAL\"\n"
-            "       \"CONSERVATIVE\""},
+                    {"conf_target", RPCArg::Type::NUM, /* default */ "wallet default", "Confirmation target (in blocks), or fee rate (for " + CURRENCY_UNIT + "/KB or " + ToUpper(CURRENCY_ATOM) + "/B estimate modes)"},
+                    {"estimate_mode", RPCArg::Type::STR, /* default */ "UNSET", std::string() + "The fee estimate mode, must be one of (case insensitive):\n"
```
The `" (case insensitive)"` addition is part of the wrong commit.

kallewoof replied:
Moved to 'make case insensitive' commit.

promag replied:
874b40cea1644812a416a517c3a4a3ab1db89144

What's the motivation to support case insensitive?

luke-jr commented on src/util/fees.cpp:
```diff
@@ -26,16 +30,35 @@ std::string StringForFeeReason(FeeReason reason) {
+}
 
-    if (mode == fee_modes.end()) return false;
+std::string FeeModes(const std::string& delimiter = ", ") {
```
Commit message calls this `FeeMode` - probably should be `FeeModes`, so fix commit message

kallewoof commented:
@luke-jr Thanks for feedback, addressed!

Sjors commented:
reACK eda3439

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3323,16 +3323,20 @@ static UniValue bumpfee(const JSONRPCRequest& request)
         RPCTypeCheckObj(options,
             {
                 {"confTarget", UniValueType(UniValue::VNUM)},
+                {"conf_target", UniValueType(UniValue::VNUM)},
```
Why is this adding a new key without documentation and whose value is silently ignored if `confTarget` is given? Doesn't make any sense to me and seems completely unrelated to this pull request anyway.

Sjors replied:
Looks like this could be avoided entirely. But at least there should be a throw if both are set. You could also rename the documented options to `conf_target`, but keep `confTarget` as an undocumented alias for backwards compatibility. That should be in the release note though.

kallewoof replied:
Documentation updated/release notes added. Also throws if both are given.

MarcoFalke commented on doc/release-notes-11413.md:
```diff
@@ -0,0 +1,8 @@
+Low-level RPC changes
```
Should be "updated or changed RPC", as low level changes are meant for things that are irrelevant to users

kallewoof replied:
Switched to "Updated or changed RPC".

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -813,10 +812,9 @@ static UniValue sendmany(const JSONRPCRequest& request)
-            "       \"CONSERVATIVE\""},
+                    {"estimate_mode", RPCArg::Type::STR, /* default */ "UNSET", std::string() + "The fee estimate mode, must be one of:\n"
+            "       \"" + FeeModes("\"\n"
+            "       \"") + "\""},
```
You don't need to pass in any whitespace here. Whitespace is done by RPCHelpMan

kallewoof replied:
Point. Fixed.

MarcoFalke commented on src/util/fees.cpp:
```diff
@@ -28,8 +30,8 @@ std::string StringForFeeReason(FeeReason reason) {
-const std::map<std::string, FeeEstimateMode>& FeeModeMap() {
-    static const std::map<std::string, FeeEstimateMode> fee_modes = {
+const std::vector<std::pair<std::string, FeeEstimateMode>>& FeeModeMap() {
+    static const std::vector<std::pair<std::string, FeeEstimateMode>> fee_modes = {
```
could squash this commit (eda34393d51c59ace68e08df2d6e00870fe5d797)?

kallewoof replied:
Squashed.

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -158,6 +158,30 @@ static std::string LabelFromValue(const UniValue& value)
     return label;
 }
 
+static void SetFeeEstimateMode(CWallet* const pwallet, CCoinControl& cc, const UniValue& estimate_mode, const UniValue& estimate_param)
```
style-nit: pwallet can be a reference, since it is non-nullable?

kallewoof replied:
You mean `const CWallet& wallet` instead? I don't see that used very often anywhere, and it seems this thing is usually a `std::shared_ptr`.

MarcoFalke replied:
Heh, I'll just link to #15922

MarcoFalke commented:
Concept ACK. Some questions, which you are free to ignore.

kallewoof committed:
fees: add FeeModes doc helper function

MarcoFalke commented on src/util/fees.cpp:
```diff
@@ -26,16 +30,35 @@ std::string StringForFeeReason(FeeReason reason) {
+}
 
-    if (mode == fee_modes.end()) return false;
+std::string FeeModes(const std::string& delimiter = ", ") {
```
Sorry to keep nit-picking this, but you set a delimiter every time this function is called. But then you set a default value in the header, as well as in the cpp file. At the very least the default should be removed from the cpp file. I'd argue to remove it in all places.

kallewoof replied:
A default `", "` seemed like a sensible thing, but you're right that I don't use the default anywhere.

Also, huh. I thought re-declaring the defaults in the .cpp file would result in a compiler error. Regardless, removed both.

Sjors commented:
Code review re-ACK c109001c9bb486eff3112d095a4639626273becc (and help text still looks correct)

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3323,16 +3323,23 @@ static UniValue bumpfee(const JSONRPCRequest& request)
             true, true);
 
-        if (options.exists("confTarget") && options.exists("totalFee")) {
-            throw JSONRPCError(RPC_INVALID_PARAMETER, "confTarget and totalFee options should not both be set. Please provide either a confirmation target for fee estimation or an explicit total fee for the transaction.");
```
1c0d97daa09f2fc792b8675ea535ba03c8a7b86f

There was no test for this 😞 And the new errors don't have tests either. Care to add?

kallewoof replied:
Hello.

kallewoof replied:
Added tests for invalid parameters.

promag commented on src/util/fees.cpp:
```diff
@@ -26,16 +29,32 @@ std::string StringForFeeReason(FeeReason reason) {
-    fee_estimate_mode = mode->second;
-    return true;
+bool FeeModeFromString(const std::string& mode_string, FeeEstimateMode& fee_estimate_mode) {
+    auto vector = FeeModeMap();
```
457396a4091bb3c27b1e982fb7e08442120b15df

Why?

kallewoof replied:
You mean why not just do `: FeeModeMap()` directly? Good point.

promag commented on src/util/fees.cpp:
```diff
@@ -26,16 +29,32 @@ std::string StringForFeeReason(FeeReason reason) {
 
-bool FeeModeFromString(const std::string& mode_string, FeeEstimateMode& fee_estimate_mode) {
-    static const std::map<std::string, FeeEstimateMode> fee_modes = {
+const std::vector<std::pair<std::string, FeeEstimateMode>>& FeeModeMap() {
```
457396a4091bb3c27b1e982fb7e08442120b15df

Why have this as a function? Could be a static variable in this translation unit, like
```cpp
static const std::vector<std::pair<std::string, FeeEstimateMode>> FEE_MODES{
    {"UNSET", FeeEstimateMode::UNSET},
    {"ECONOMICAL", FeeEstimateMode::ECONOMICAL},
    {"CONSERVATIVE", FeeEstimateMode::CONSERVATIVE},
    {(CURRENCY_UNIT + "/KB"), FeeEstimateMode::BTC_KB},
    {(ToUpper(CURRENCY_ATOM) + "/B"), FeeEstimateMode::SAT_B},
};
```

kallewoof replied:
Windows build breaks if I do this, unfortunately.

kallewoof replied:
Now that `ToUpper` is no longer used, this may actually work in Windows; tentatively removing the function wrapper.

promag commented on src/util/fees.cpp:
```diff
@@ -26,16 +29,32 @@ std::string StringForFeeReason(FeeReason reason) {
 
-    if (mode == fee_modes.end()) return false;
+std::string FeeModes(const std::string& delimiter) {
+    std::ostringstream r;
```
457396a4091bb3c27b1e982fb7e08442120b15df

Is it really necessary to `#include <sstream>` just to concat a couple of strings? Could just do
```cpp
for (i : ...) {
    if (!res.empty()) res += delimiter;
    res += i.first
}
```

kallewoof replied:
I don't really see the harm in including it (it won't make the binary bigger or anything) but it's such a trivial set of operations that I doubt it'll matter in the long run.

I.e. removing the include and just stacking onto an `std::string`.

promag commented on src/util/fees.cpp:
```diff
@@ -26,16 +29,32 @@ std::string StringForFeeReason(FeeReason reason) {
+}
 
-    if (mode == fee_modes.end()) return false;
+std::string FeeModes(const std::string& delimiter) {
```
457396a4091bb3c27b1e982fb7e08442120b15df

This is a utility function for the RPC help, I'd move to `rpcwallet.cpp` and make it static there.

BTW, according to dev notes

> Braces on new lines for classes, functions, methods.

kallewoof replied:
Fixed style.

kallewoof replied:
Forgot to respond to the first point; I could move it, but it would mean having to make `fee_modes` available to the public, which doesn't seem worth it.

Sjors commented:
@promag it's case insensitive to avoid bike-shedding "sat/B" vs. "SAT/B" and "BTC/kB" vs "BTC/KB"

promag commented on src/util/fees.cpp:
```diff
@@ -26,16 +30,35 @@ std::string StringForFeeReason(FeeReason reason) {
+    auto searchkey = ToUpper(mode_string);
+    auto vector = FeeModeMap();
+    for (const auto& pair : vector) {
+        if (pair.first == searchkey) {
```
Regarding https://github.com/bitcoin/bitcoin/pull/11413/files#r313929142

Then I'd do `ToUpper(pair.first) == searchkey)` and display the right casing.

kallewoof replied:
Done.

Sjors commented:
Mysterious Windows failure:
<img width="825" alt="Schermafbeelding 2019-08-28 om 10 02 31" src="https://user-images.githubusercontent.com/10217/63837028-faa7ba00-c97a-11e9-8f79-369da0ebd10f.png">

kallewoof commented:
@Sjors Ah.. yeah, I think I got that before as well. For some reason it fails to up-case..

promag commented on src/util/fees.cpp:
```diff
@@ -26,16 +30,33 @@ std::string StringForFeeReason(FeeReason reason) {
-    auto mode = fee_modes.find(mode_string);
-
-    if (mode == fee_modes.end()) return false;
+static const std::vector<std::pair<std::string, FeeEstimateMode>> fee_modes = {
```
Should be `FEE_MODES`.

promag commented on test/functional/wallet_bumpfee.py:
```diff
@@ -87,6 +88,38 @@ def run_test(self):
+def test_invalid_parameters(node, dest_address):
+    txid = spend_one_input(node, dest_address)
+    # invalid estimate mode
+    assert_raises_rpc_error(-8, "Invalid estimate_mode parameter", node.bumpfee, txid, {
```
nit, space after `:` here and below.

promag commented on doc/release-notes-11413.md:
```diff
@@ -0,0 +1,10 @@
+----------------------
+
+The `bumpfee`, `fundrawtransaction`, `sendmany`, `sendtoaddress`, and `walletcreatefundedpsbt`
+RPC commands have been updated to include two new fee estimation methods "BTC/KB" and "SAT/B".
```
fix case.

promag commented on doc/release-notes-11413.md:
```diff
@@ -0,0 +1,10 @@
+
+In addition, the estimate_mode parameter is now case insensitive for all of the above RPC commands.
+
+The `bumpfee` command now uses `conf_target` rather than `confTarget` in the options.
```
Should mention that `confTarget` is still supported but clients should transition to `conf_target`.

kallewoof replied:
I don't really see the point. Literally nobody will be affected by this (i.e. they will either see the release notes and switch, or not see them and not notice the change).

promag commented on src/util/fees.cpp:
```diff
@@ -26,16 +30,33 @@ std::string StringForFeeReason(FeeReason reason) {
+{
+    std::string r;
+    std::string prefix = "";
+    for (const auto& i : fee_modes) {
```
s/i/pair like below?

kallewoof commented:
Yikes... turning back into a function. @promag 

<img width="1003" alt="Screen Shot 2019-08-28 at 23 49 33" src="https://user-images.githubusercontent.com/250224/63866672-89095380-c9ee-11e9-9b0d-a5850ba48d6b.png">

promag commented:
@kallewoof well..

kallewoof committed:
tests for bumpfee / estimate_modes

* invalid parameter tests for bumpfee
* add tests for no conf_target explicit estimate_modes

kallewoof commented:
@promag Sorry, screwed up rebase; reverted now.

kallewoof commented:
I found a different workaround; apparently MSVC does not bug out if you put the `const std::string`s in the `.h` file. We do that elsewhere so I assume nobody will complain about moving `CURRENCY_UNIT` into `feerate.h` from `feerate.cpp`.

promag reviewed (commented):
Shouldn't `conf_target` be required when using an explicit estimate mode?

kallewoof commented:
@promag It should indeed. I added tests and fixed this.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -158,6 +158,33 @@ static std::string LabelFromValue(const UniValue& value)
+        cc.m_feerate = CFeeRate(fee_rate);
+        // default RBF to true for explicit fee rate modes
+        if (cc.m_signal_bip125_rbf == boost::none) cc.m_signal_bip125_rbf = true;
+    } else {
```
`else if (! estimate_param.isNull()) {`?

kallewoof replied:
Done

instagibbs commented:
concept ACK, but I am *really* wary of the implementation of the arguments.

The meaning of `conf_target` basically inverts with this new argument. Consider:
`sendtoaddress 2N7GTLzhS9n3FzwAQohAgnk2kGvt8eE9nng 1 "" "" false true 1008`

and

`sendtoaddress 2N7GTLzhS9n3FzwAQohAgnk2kGvt8eE9nng 1 "" "" false true 1008 sat/B`

and the opposite way:
`1` vs `1 sat/B` is likely considerable.

I'd rather we "overload" the `conf_target` and parse `30sat/B` directly, disallowing estimate mode to be specified maybe other than `unset`.

instagibbs commented on test/functional/wallet_basic.py:
```diff
@@ -338,6 +373,52 @@ def run_test(self):
+        postbalance = self.nodes[2].getbalance()
+        fee = prebalance - postbalance - Decimal('1')
+        assert_fee_amount(fee, tx_size, Decimal('0.00002000'))
+
```
Add a test for negative/too low feerates?

kallewoof replied:
Done

Sjors commented:
> I'd rather we "overload" the conf_target and parse 30sat/B directly, disallowing estimate mode to be specified maybe other than unset.

If that's easy to implement I also think it's significantly more clear.

instagibbs commented:
I think it's significantly safer. The errors would end up being someone intended to add a string after the number, and not, and it ending up being converted into a potentially much lower feerate, I think? That is salvageable usually.

instagibbs commented:
The issue still remains for someone who types `1` and forgets to write `sat/B`. I'm not sure if there's a backwards compatible way to solve this.

luke-jr commented:
Specifying an amount as a string like that is completely foreign to all existing RPC interfaces.

The number is already interpreted differently based on the mode string.

luke-jr commented:
Maybe `conf_target` as a named param should reject explicit modes, and it should be specified as another named param for explicit?

That might require #11660

kallewoof commented:
@instagibbs Yeah, typing '1' would mean 'confirm within 1 block' which is directly opposite of '1 sat/b'.

Adding number suffixes only partially solves the problem. Someone who accidentally hits enter forgetting the 'sat/b' suffix will have exactly the same issue.

Perhaps the explicit modes should not overload `conf_target` but instead use a new argument entirely. Alternatively we make the estimate_mode include the actual number, so people will do

```
[...]conf_target=null estimate_mode=1sat/b [...]
```

which means the ambiguity is removed. If they do `estimate_mode=1` it will error. And `conf_target` no longer both means "low value = high priority = potentially high fees" and "low value = low fees" at the same time, depending on the estimate mode.

kallewoof commented:
Another alternative is to actually require suffixes for all conf_targets, but that is a compatibility breach (1 block, 1 sat/b, 1 btc/kb).

Sjors commented:
I don't like including the string in the same param (without a space). Nor do I like the verboseness of named params.

Making a breaking change might be worth it for the extra safety. It would fail quite gracefully, i.e. it just stops sending. The impact can be reduced by not requiring `estimate_mode` when using named parameters.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -175,6 +175,36 @@ static std::string LabelFromValue(const UniValue& value)
+        }
+        cc.m_feerate = CFeeRate(fee_rate);
+        // if feerate < 0.1 sat/b, throw
+        if (fee_rate < 100) {
```
This shouldn't be hard-coded

kallewoof replied:
Turned into constant. Also did same for the conversion a few lines above.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -39,6 +39,10 @@
 
+static const uint32_t WALLET_MIN_FEERATE = 100; // sat/b
+
+static const uint32_t WALLET_BTC_KB_TO_SAT_B = 100000; // 1 sat / B = 0.00001 BTC / kB
```
Please use the `COIN` const for this...

kallewoof replied:
Changed to `COIN / 1000`.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -39,6 +39,10 @@
 
 static const std::string WALLET_ENDPOINT_BASE = "/wallet/";
 
+static const uint32_t WALLET_MIN_FEERATE = 100; // sat/b
```
This is wrong. Relay min feerate is already defined elsewhere, and user-configurable.

kallewoof replied:
Using `DEFAULT_MIN_RELAY_TX_FEE` instead.

kallewoof replied:
Actually, I would have to include `validation.h` to access that, and I can't access it via `extern` because it's `static`. I'm just doing `1000 /* DEFAULT_MIN_RELAY_TX_FEE */` for now.

luke-jr replied:
Using the default would be wrong too. It should use the configured policy or current mempool limit...

kallewoof replied:
I don't see a problem with someone making a tx that is not relayable at this point in time (but relayable when tx traffic drops). In fact, I'm not even sure this check is necessary in the first place.

luke-jr replied:
In that case, it should be the *configured* relay minimum policy (or no check at all).

kallewoof replied:
I'm honestly leaning towards removing this check. Thanks for the nudge.

Sjors commented:
Tested ACK 398be1c modulo squash.

Positional arguments are nigh-impossible to use, but let's worry about that later...
```
bitcoin-cli sendtoaddress bc... 1 "" "" false true 1 sat/b
bitcoin-cli sendmany "" '{"bc...": 1}' 0 "" '[]' true 1 sat/b
```

A mix of positional and named kind of works:
```
sendtoaddress bcrt1qgr3zfmvqzl4uv3utz6h849x0fu5a8mvg695jde 1 -conf_target=1 estimate_mode=sat/b
```

luke-jr commented on test/functional/wallet_basic.py:
```diff
@@ -338,6 +391,74 @@ def run_test(self):
+        # send with explicit btc/kb fee
+        self.sync_all(self.nodes[0:3])
+        self.log.info("test explicit fee (sendtoaddress as btc/kb)")
+        self.nodes[0].generate(1)
```
The `sync_all` needs to be moved after `generate`, or `prebalance` could end up with a stale balance and cause the check below (line 426) to fail.

kallewoof replied:
Moved.

Sjors commented:
Code review re-ACK 9721534f1c929ce275ac0516d5777bc1063f7495

Sjors commented:
Still works when rebased. @achow101 @instagibbs maybe you can (re-)review?

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3341,7 +3341,7 @@ static UniValue bumpfee(const JSONRPCRequest& request)
                     {"options", RPCArg::Type::OBJ, RPCArg::Optional::OMITTED_NAMED_ARG, "",
                         {
-                            {"confTarget", RPCArg::Type::NUM, /* default */ "wallet default", "Confirmation target (in blocks)"},
+                            {"conf_target", RPCArg::Type::NUM, /* default */ "wallet default", "Confirmation target (in blocks)"},
```
convert tests for coverage?

kallewoof replied:
I thought I did -- which one did I miss?

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -183,6 +185,33 @@ static std::string LabelFromValue(const UniValue& value)
+
+    if (cc.m_fee_mode == FeeEstimateMode::BTC_KB || cc.m_fee_mode == FeeEstimateMode::SAT_B) {
+        if (estimate_param.isNull()) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Selected estimate_mode requires a confirmation target");
```
`s/confirmation target/fee rate/` ?

kallewoof replied:
Right! Thanks.

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -183,6 +185,33 @@ static std::string LabelFromValue(const UniValue& value)
     return label;
 }
 
+static void SetFeeEstimateMode(CWallet* const pwallet, CCoinControl& cc, const UniValue& estimate_mode, const UniValue& estimate_param)
```
A bit of documentation would go a long way for function declaration.

kallewoof replied:
Added documentation!

instagibbs commented on doc/release-notes-11413.md:
```diff
@@ -0,0 +1,10 @@
+Updated or changed RPC
+----------------------
+
+The `bumpfee`, `fundrawtransaction`, `sendmany`, `sendtoaddress`, and `walletcreatefundedpsbt`
```
Note that this will also trigger bip125 opt-in.

kallewoof replied:
Added node.

instagibbs reviewed (commented):
looks good, couple of notes

kallewoof commented:
@instagibbs Thanks for the review! Addressed everything except the coverage part which I am hoping you'll clarify.

Sjors commented:
Code review re-ACK 74f9d2797bd199fcd4cfc30e490bbac6224424f9

kallewoof commented:
@Sjors Nice, good catch! Where in wallet_basic did you add that test, btw?

kallewoof commented:
Updated code to now return an error when creating a transaction if the requested explicit fee rate differs from the required fee rate. This prevents silently ignoring the user when they want a fee rate < 1 sat/b -- it now gives an error, as you would expect.

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -2660,6 +2660,12 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
+            // Do not, ever, assume that it's fine to change the fee rate if the user has explicitly
+            // provided one
+            if (coin_control.m_feerate && nFeeRateNeeded != *coin_control.m_feerate) {
+                strFailReason = "Fee rate needed differs from explicitly provided fee rate";
```
"needed" is rather vague. I assume this is based on the mempool relay fee? So maybe: `Fee rate is too low for the current mempool. Increase it to %f sat/b or modify -maxmempool`

kallewoof replied:
That's a bit weird, though, for the case when the relay fee is at the minimum. We can't really ask people to change `minrelaytxfee` either, as that would mean their node ends up with non-relayable (to most other nodes) transactions. I don't really see a better alternative though, so I'm going with your version with some tweaks:

I built on this a tiny bit, adding a new optional `FeeEstimateMode` parameter to `CFeeRate::ToString()`, which returns sat/b formatted fee rate if estimate mode is sat/b, otherwise the (pre-existing) btc/kb fee rate. So if a user provides an explicit fee rate using `sat/b`, they get the needed one back in sat/b, otherwise it is expressed in btc/kb.

Sjors replied:
There have been occasional rumblings of allowing a lower relay fee than 1 sat/byte, so it's nice to future proof. It's not an ideal message, but definitely better than silently changing the input.

kallewoof replied:
This should cover for the <1sat/b case as well. All the way down to 0.001 sat/b.

kallewoof replied:
I updated the above comment a bit to clarify that 1 sat/b is not a hardcoded thing.

Sjors replied:
IIUC coin selection uses `nFeeRateNeeded` here:
```cpp
coin_selection_params.effective_fee = nFeeRateNeeded;
```

This doesn't make sense for transaction with a future nLockTime, and with PSBTs it's debatable, since we don't know when the user intends to broadcast. For `sendmany` and `sendtoaddress` the current behavior is fine, but those don't support either.

For PSBT I'm inclined to think that `effective_fee` should be set to the manual fee. In #16378 I plan to do the same for the new `send` RPC. It returns the hex in that case (as it does with `nLockTime`).

kallewoof replied:
So basically, you'd prefer if I could force a transaction using a fee rate that is currently not actually accepted by the mempool? I initially did that, but the result is that the tx goes into the mempool but is rejected by the node's peers, so it is never mined (unless the node itself generates a block). That seems awfully confusing.

Perhaps there should be an exception for the locktime>now case, where a warning is produced but the tx is created since it can't go into the mempool anyway, due to locktime not unfulfilled.

This can be achieved by setting the `fOverrideFeeRate` to `true`, btw.

Sjors replied:
I think the current error is fine when the user intends to broadcast immediately. Only when that's not clearly the case, e.g. for the PSBT methods, we should still allow transaction creation.

Sjors replied:
`fOverrideFeeRate` could do the trick, but it's pretty ugly: it basically "sabotages" `GetMinimumFeeRate`. It seems more readable to skip `GetMinimumFeeRate` altogether when you want to ignore it.

kallewoof committed:
MOVEONLY: Make FeeEstimateMode available to CFeeRate

Can verify move-only with:

    git log -p -n1 --color-moved

This commit is move-only and doesn't change code or affect behavior.

kallewoof committed:
rpc/wallet: add two explicit modes to estimate_mode

kallewoof committed:
policy: optional FeeEstimateMode param to CFeeRate::ToString

luke-jr commented on src/policy/feerate.cpp:
```diff
@@ -37,7 +35,10 @@ CAmount CFeeRate::GetFee(size_t nBytes_) const
 {
-    return strprintf("%d.%08d %s/kB", nSatoshisPerK / COIN, nSatoshisPerK % COIN, CURRENCY_UNIT);
+    switch (fee_estimate_mode) {
+    case FeeEstimateMode::SAT_B:  return strprintf("%d.%03d %s/b", nSatoshisPerK / 1000, nSatoshisPerK % 1000, CURRENCY_ATOM);
```
Satoshis per bit?

kallewoof replied:
Good catch, fixed!

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -2583,6 +2583,7 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
                          int& nChangePosInOut, std::string& strFailReason, const CCoinControl& coin_control, bool sign)
 {
     CAmount nValue = 0;
+    nFeeRet = 0;
```
is this valuable by itself?

kallewoof replied:
Probably. I had a garbage value being returned and it ended up telling me "you need at least 4587368374.13587638 bitcoin in fees!" here:

https://github.com/bitcoin/bitcoin/blob/3f826598a42dcc707b58224e94c394e30a42ceee/src/wallet/rpcwallet.cpp#L353-L357

kallewoof replied:
PR-ified as #18274.

instagibbs reviewed (approved):
utACK https://github.com/bitcoin/bitcoin/pull/11413/commits/021c253961d6c51dbc1d031449b1b3d0f5edef8b

I'll hopefully get around to testing a bit.

instagibbs commented:
something to think about out loud for the future: If we want to support implicit CPFP, this new mode would probably turn that off, at least by default.

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -2660,6 +2660,12 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
+            // Do not, ever, assume that it's fine to change the fee rate if the user has explicitly
+            // provided one
+            if (coin_control.m_feerate && nFeeRateNeeded > *coin_control.m_feerate) {
+                strFailReason = strprintf("Fee rate is too low for the current mempool. Increase it to %s or modify -maxmempool", nFeeRateNeeded.ToString(coin_control.m_fee_mode));
```
Actually I find this `-maxmempool` advice pretty dangerous for users. It's going to simply result in stuck transactions, no?

kallewoof replied:
I think it might get stuck for the user themselves, but unless their peers also increase their mempool sizes, it won't propagate very far. But I'm now a bit hesitant to suggest using it as it'll probably not do what people expect. (As you say, it will get stuck in the sense that it won't get mined or propagate properly..)

Sjors replied:
That's something we can point out in `-maxmempool` documentation. Once #18038 lands the mempool will at least hold on to the transaction if you unload a wallet.

sipa replied:
Agree that we shouldn't mention -maxmempool here; it indeed likely will not do what users expect it to do.

kallewoof replied:
Changed to "Fee rate is too low for the current mempool. Increase it to %s or wait."

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -2660,6 +2660,12 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
 
             // Get the fee rate to use effective values in coin selection
             CFeeRate nFeeRateNeeded = GetMinimumFeeRate(*this, coin_control, &feeCalc);
+            // Do not, ever, assume that it's fine to change the fee rate if the user has explicitly
```
Yeah, it could be its own PR. Maybe I should split these out..

kallewoof replied:
PR-ified in #18275.

Sjors commented:
re-ACK 8326a544b7839e362c7e530bebfffb02ba1ddb34 modulo rebase for #18274 (which contains 7864c1f)

kallewoof commented:
Good catch; dropped 7864c1f.

Sjors commented:
utACK 04e94beb1d9f22809b64bda2c7e92c1e911d2227

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/11413/commits/04e94beb1d9f22809b64bda2c7e92c1e911d2227

sipa reviewed (commented):
utACK 04e94beb1d9f22809b64bda2c7e92c1e911d2227, apart from the -maxmempool nit below.

kallewoof commented:
Addressed @sipa comments. Please re-review! :)

Sjors commented:
re-utACK 6fef8dcd771f1a628d2c4bee2682a98312bcd80b

Sjors reviewed (approved):
re-utACK 6fef8dcd771f1a628d2c4bee2682a98312bcd80b: recommending patience can't hurt :-)

instagibbs commented:
re-utACK https://github.com/bitcoin/bitcoin/pull/11413/commits/6fef8dcd771f1a628d2c4bee2682a98312bcd80b

kallewoof commented:
Rebased

Sjors commented:
re-utACK a29b9939b6537e60a9dd47e0ae49968851cac251

Sjors commented:
re-utACK 9afd897, just a rebase and some tests have moved.

luke-jr commented on test/functional/wallet_basic.py:
```diff
@@ -350,6 +403,87 @@ def run_test(self):
+            fee = prebalance - postbalance - Decimal('1')
+            assert_fee_amount(fee, tx_size, Decimal('0.00002000'))
+
+            # Import address and private key to check correct behavior of spendable unspents
+            # 1. Send some coins to generate new UTXO
+            address_to_import = self.nodes[2].getnewaddress()
+            txid = self.nodes[0].sendtoaddress(address_to_import, 1)
+            self.nodes[0].generate(1)
+            self.sync_all(self.nodes[0:3])
+
```
I believe this was a mis-rebase. This test goes back to 2016, and is above the added tests (prior to the rebase, it was here, below the added tests)

kallewoof replied:
Yes, that is correct! Good catch. Fixed.

kallewoof commented:
Dropped 'fee rate changed' commit, as it is its own PR now (#18275).

luke-jr commented on src/util/fees.cpp:
```diff
@@ -29,16 +33,33 @@ std::string StringForFeeReason(FeeReason reason) {
+    {"unset", FeeEstimateMode::UNSET},
+    {"economical", FeeEstimateMode::ECONOMICAL},
+    {"conservative", FeeEstimateMode::CONSERVATIVE},
+    {(CURRENCY_UNIT + "/kB"), FeeEstimateMode::BTC_KB},
+    {(CURRENCY_ATOM + "/B"), FeeEstimateMode::SAT_B},
```
We can't use `CURRENCY_*` here in global scope, since the initialisation order is  undefined. A simple static function wrapper should work.

kallewoof replied:
Odd that no compiler warning or anything was emitted about it. I've updated the code to wrap the map.

Sjors replied:
For future reference, the original: 3bad9a438769c6749063ab5ab405a62f6efea8cc

jonatack commented:
Concept ACK

Sjors commented:
re-tACK a9ad35d3016d9d912472861895b8f5160e24cc12

kallewoof commented:
Rebased to see if weird FreeBSD error would go away. No changes since a9ad35d

jonatack commented on test/functional/wallet_bumpfee.py:
```diff
@@ -92,6 +93,38 @@ def run_test(self):
+    assert_raises_rpc_error(-3, "Unexpected key totalFee", node.bumpfee, txid, {
+        "conf_target": 123,
+        "totalFee": 456,
+    })
```
69823cc can omit lines 118-127 as the `totalFee` argument is now removed

kallewoof replied:
Thanks, done!

jonatack commented on doc/release-notes-11413.md:
```diff
@@ -0,0 +1,11 @@
+The target is the fee expressed explicitly in the given form. Note that use of this feature
+will trigger BIP 125 (replace-by-fee) opt-in.
+
+In addition, the estimate_mode parameter is now case insensitive for all of the above RPC commands.
```
5ee0e46 nit: s/estimate_mode/`estimate_mode`/

jonatack commented on doc/release-notes-11413.md:
```diff
@@ -0,0 +1,11 @@
+----------------------
+
+The `bumpfee`, `fundrawtransaction`, `sendmany`, `sendtoaddress`, and `walletcreatefundedpsbt`
+RPC commands have been updated to include two new fee estimation methods "BTC/kB" and "sat/B".
```
5ee0e46 nit: missing comma after "methods"

kallewoof replied:
I think it's okay as it is. The methods `a` and `b` were added.

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -385,6 +421,8 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
             + HelpExampleCli("sendtoaddress", "\"" + EXAMPLE_ADDRESS[0] + "\" 0.1 \"\" \"\" true")
             + HelpExampleRpc("sendtoaddress", "\"" + EXAMPLE_ADDRESS[0] + "\", 0.1, \"donation\", \"seans outpost\"")
+            + HelpExampleCli("sendtoaddress", "\"" + EXAMPLE_ADDRESS[0] + "\" 0.1 \"\" \"\" false true 0.00002 " + (CURRENCY_UNIT + "/kB"))
+            + HelpExampleCli("sendtoaddress", "\"" + EXAMPLE_ADDRESS[0] + "\" 0.1 \"\" \"\" false true 2 " + (CURRENCY_ATOM + "/B"))
```
0fb63f0 these two lines should be placed before the `HelpExampleRpc` in line 423, as currently the help prints:

```
Examples:
> bitcoin-cli sendtoaddress "bc1q09vm5lfy0j5reeulh4x5752q25uqqvz34hufdl" 0.1
> bitcoin-cli sendtoaddress "bc1q09vm5lfy0j5reeulh4x5752q25uqqvz34hufdl" 0.1 "donation" "seans outpost"
> bitcoin-cli sendtoaddress "bc1q09vm5lfy0j5reeulh4x5752q25uqqvz34hufdl" 0.1 "" "" true
> curl --user myusername --data-binary '{"jsonrpc": "1.0", "id": "curltest", "method": "sendtoaddress", "params": ["bc1q09vm5lfy0j5reeulh4x5752q25uqqvz34hufdl", 0.1, "donation", "seans outpost"]}' -H 'content-type: text/plain;' http://127.0.0.1:8332/
> bitcoin-cli sendtoaddress "bc1q09vm5lfy0j5reeulh4x5752q25uqqvz34hufdl" 0.1 "" "" false true 0.00002 BTC/kB
> bitcoin-cli sendtoaddress "bc1q09vm5lfy0j5reeulh4x5752q25uqqvz34hufdl" 0.1 "" "" false true 2 sat/B
```

kallewoof replied:
Good catch, fixed

jonatack reviewed (commented):
ACK 5ee0e46b93b2ca41eda9ee46fa59c6834689f0d5

Kudos on your patience on this useful feature. Feel free to ignore the comments below. The RPC examples could be updated (as you've done for sendtoaddress) and the tests could use some cleanup and refactoring, but nothing that can't be done in a follow-up. Happy to re-ACK if you retouch.

kallewoof commented on src/util/fees.cpp:
```diff
@@ -29,16 +33,37 @@ std::string StringForFeeReason(FeeReason reason) {
+}
 
-    if (mode == fee_modes.end()) return false;
+std::string FeeModes(const std::string& delimiter)
```
Seems like an odd default, tbh, even though it would reduce code elsewhere.

kallewoof commented on test/functional/wallet_basic.py:
```diff
@@ -219,6 +219,59 @@ def run_test(self):
+        self.nodes[2].generate(1)
+        self.sync_all(self.nodes[0:3])
+        node_2_bal = self.check_fee_amount(self.nodes[2].getbalance(), node_2_bal - Decimal('10'), explicit_fee_per_byte, self.get_vsize(self.nodes[2].gettransaction(txid)['hex']))
+        assert_equal(self.nodes[2].getbalance(), node_2_bal)
```
Done.

jonatack reviewed (commented):
ACK 25dac9fa65243ca8db02df2 I think this should be merged after all this time, even though it looks to me like there are needed follow-ups, fixes and test coverage to be added (see further down), which I don't mind helping out with, if wanted.

-----
Summary of the documentation of the changed RPCs:

`bumpfee`
- fee_rate option: "feerate (NOT total fee) to pay, in BTC per kB"

`fundrawtransaction`
- feeRate option: "Set a specific fee rate in BTC/kB"
- fee rate set in conf_target (for BTC/kB or sat/B modes)

`sendmany`
- fee rate set in conf_target: "fee rate (for BTC/kB or sat/B estimate modes)"

`sendtoaddress`
- fee rate set in conf_target: "fee rate (for BTC/kB or sat/B estimate modes)"

`walletcreatefundedpsbt`
- feeRate option: "Set a specific fee rate in BTC/kB"
-----
Thoughts:

- I find this pretty confusing and worth harmonizing.

- `walletcreatefundedpsbt` does not work with `feeRate` + an explicit estimate mode. Not a surprise given the logic in `FundTransaction`, wallet/rpcwallet.cpp::3005. Here are added tests in `rpc_psbt.py` from line 154:

```python
        # Existing test.
        # feeRate of 0.1 BTC / KB produces a total fee slightly below -maxtxfee (~0.05280000):
        res = self.nodes[1].walletcreatefundedpsbt([{"txid":txid,"vout":p2wpkh_pos},{"txid":txid,"vout":p2sh_p2wpkh_pos},{"txid":txid,"vout":p2pkh_pos}], {self.nodes[1].getnewaddress():29.99}, 0, {"feeRate": 0.1})
        assert_greater_than(res["fee"], 0.05)
        assert_greater_than(0.06, res["fee"])

        # Same test using conf_target and estimate_mode btc/kb is green.
        res = self.nodes[1].walletcreatefundedpsbt(
            [{"txid":txid,"vout":p2wpkh_pos}, {"txid":txid,"vout":p2sh_p2wpkh_pos}, {"txid":txid,"vout":p2pkh_pos}], {self.nodes[1].getnewaddress():29.99}, 0,
            {
                "conf_target": 0.1,
                "estimate_mode": 'btc/kb',
            }
        )
        assert_greater_than(res["fee"], 0.05)
        assert_greater_than(0.06, res["fee"])

        # Same test using conf_target and estimate_mode sat/b is green.
        res = self.nodes[1].walletcreatefundedpsbt(
            [{"txid":txid,"vout":p2wpkh_pos}, {"txid":txid,"vout":p2sh_p2wpkh_pos}, {"txid":txid,"vout":p2pkh_pos}], {self.nodes[1].getnewaddress():29.99}, 0,
            {
                "conf_target": 10000,
                "estimate_mode": 'sat/b',
            }
        )
        assert_greater_than(res["fee"], 0.05)
        assert_greater_than(0.06, res["fee"])

        # But the same test using feeRate and estimate_mode btc/kb or
        # sat/b fails with "Cannot specify both estimate_mode and feeRate".
        res = self.nodes[1].walletcreatefundedpsbt(
            [{"txid":txid,"vout":p2wpkh_pos}, {"txid":txid,"vout":p2sh_p2wpkh_pos}, {"txid":txid,"vout":p2pkh_pos}],
            [{"txid":txid,"vout":p2wpkh_pos}, {"txid":txid,"vout":p2sh_p2wpkh_pos}, {"txid":txid,"vout":p2pkh_pos}], {self.nodes[1].getnewaddress():29.99}, 0,
            {
                "feeRate": 10000,
                "estimate_mode": 'sat/b',
            }
        )
```

- We need tests for `fundrawtransaction`, `walletcreatefundedpsbt` and `bumpfee`.

- We probably need to update the docs or code for `fundrawtransaction` and `walletcreatefundedpsbt` so that `feeRate` with an explicit estimation mode either works or is stated as not permitted.

jonatack commented:
If helpful, some notes I took while reviewing this PR: https://gist.github.com/jonatack/f16e51274d4254c01f682063b0f89e04

kallewoof commented:
Thanks for the detailed review @jonatack! It looks like `bumpfee` is suffering from the same issue as `walletcreatefundedpsbt`. I'm totally for merging this and doing a follow-up PR to address those, if this gets enough momentum to finally be merged after all this time. Otherwise I'll look into fixing those two.

fjahr commented on test/functional/wallet_bumpfee.py:
```diff
@@ -127,9 +150,10 @@ def test_feerate_args(self, rbf_node, peer_node, dest_address):
     assert rbfid in rbf_node.getrawmempool() and rbfid in peer_node.getrawmempool()
 
-    assert_raises_rpc_error(-8, "confTarget can't be set with fee_rate. Please provide either a confirmation target in blocks for automatic fee estimation, or an explicit fee rate.", rbf_node.bumpfee, rbfid, {"fee_rate": NORMAL, "confTarget": 1})
+    assert_raises_rpc_error(-8, "conf_target can't be set with fee_rate. Please provide either a confirmation target in blocks for automatic fee estimation, or an explicit fee rate.", rbf_node.bumpfee, rbfid, {"fee_rate": NORMAL, "confTarget": 1})
```
same below.
```suggestion
    assert_raises_rpc_error(-8, "conf_target can't be set with fee_rate. Please provide either a confirmation target in blocks for automatic fee estimation, or an explicit fee rate.", rbf_node.bumpfee, rbfid, {"fee_rate": NORMAL, "conf_target": 1})
```

fjahr reviewed (approved):
Code review ACK 25dac9fa65243ca8db02df22f484039c08114401

Just found two additional nits that can be ignored or kept for the follow-up.

Sjors commented:
re-utACK 25dac9fa65: rebased, more fancy C++, 
<img width="842" alt="Schermafbeelding 2020-06-25 om 18 37 03" src="https://user-images.githubusercontent.com/10217/85761269-e0fa0f00-b712-11ea-90c6-0f1eab891baa.png">

Strong preference for merging this and improving the rest in followups. This has been blocking #16378 for a while and rebasing/tweaking a big change can be a lot of work, also on reviewers.

jonatack commented:
Ouch, that image with the flashy white background hitting the eyes in dark mode :smile: 

> Strong preference for merging this and improving the rest in followups. This has been blocking #16378 for a while and rebasing/tweaking a big change can be a lot of work, also on reviewers.

:100:... @kallewoof has the patience of a sphinx.

MarcoFalke commented:
Sorry to be so late with this review, but it seems that this overloads the `conf_target` which used to be an integer (number of blocks) to also accept floats or strings. Then, if a float or string is passed, its unit is taken from the estimate mode (which used to be an enum string).

I believe, in the past all amounts (with a few exceptions) were passed as BTC and all feerates were passed as BTC/kB. If we start making feerates more type safe by forcing unit to be passed, then we should come up with a method that works for the whole code base and the whole RPC interface, not only for the few RPCs that happen to have two RPCArgs that can be recycled into passing a new type.
For example, in the future one could imagine an RPC to set the feefilter of a peer or to set the mempool min fee or simply `settxfee` would be updated to allow passing a type-safe feerate.
Then having multiple ways to pass a feerate would be confusing and added complexity at the call site.

Also, passing the feerate (float) as an argument that used to be an integer is going to ask for mishaps, no?

While this is obviously an improvement, and I don't want to block it, I also think that we should aim to get the RPC interface right the very first time, since it is impossible to change after release and hard to change in future releases.

MarcoFalke commented:
Oh, looks like this got merged before I left my comment.

Anyway, I'd like to hear what the reviewers think of adding a new `feerate` argument that simply passed the feerate (with unit) in one string. Then, parsing code could be shared among all places in the RPC code that use the new feerate type.

instagibbs commented:
I suggested that a long time ago but iirc people didn't like that idea...

On Thu, Jun 25, 2020, 2:00 PM MarcoFalke <notifications@github.com> wrote:

> Oh, looks like this got merged before I left my comment.
>
> Anyway, I'd like to hear what the reviewers think of adding a new feerate
> argument that simply passed the feerate (with unit) in one string. Then,
> parsing code could be shared among all places in the RPC code that use the
> new feerate type.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/11413#issuecomment-649733071>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABMAFU3BKIFRU27KYLBKMELRYOGDXANCNFSM4D43FZIA>
> .
>

laanwj commented:
> Anyway, I'd like to hear what the reviewers think of adding a new feerate argument that simply passed the feerate (with unit) in one string. Then, parsing code could be shared among all places in the RPC code that use the new feerate type.

I do agree that this would be better defined. Overloading an argument on type seems somewhat brittle, and could lead to serious accidents. Sorry for merging I thought there was agreement.

MarcoFalke commented:
Yeah, no worries. I was about to merge this myself just now when I realized that this might not play out well long-term.

jonatack commented:
I also prefer not overloading the argument. From reading the years of discussion I thought the discussion had been settled and that ship had sailed. In any case, there's work to do before it is release-ready and I'm happy to help.

MarcoFalke commented:
I am happy to review any follow-ups, so please ping me if I don't show up on my own

kallewoof commented:
@jonatack Are you up for doing a follow-up PR? I'll gladly review it if so.

jonatack commented:
@kallewoof SGTM. Fixups + tests now, then propose a universal feerate arg unless you want to jump on that.

kallewoof commented:
Go for it! :) Let me know if it becomes overwhelming and I'll see if I can pick up a part of it.

maste197 commented:
Good

Den fre 26 juni 2020 09:59kallewoof <notifications@github.com> skrev:

> Go for it! :) Let me know if it becomes overwhelming and I'll see if I can
> pick up a part of it.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/11413#issuecomment-650039097>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOTTDNHV4BKTPA3ZGB32IPLRYRIPBANCNFSM4D43FZIA>
> .
>

jonatack commented:
First follow-up is #20220.
